x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -344,11 +349,13 @@ public void clusterChanged(ClusterChangedEvent event) {
         AutoFollower(final String remoteCluster,
                      final Consumer<List<AutoFollowResult>> statsUpdater,
                      final Supplier<ClusterState> followerClusterStateSupplier,
-                     LongSupplier relativeTimeProvider) {
+                     final LongSupplier relativeTimeProvider,
+                     final Executor executor) {
             this.remoteCluster = remoteCluster;
             this.statsUpdater = statsUpdater;
             this.followerClusterStateSupplier = followerClusterStateSupplier;
             this.relativeTimeProvider = relativeTimeProvider;
+            this.executor = Objects.requireNonNull(executor);
         }
 
         void start() {
```

Comment:
Can we assert that method `start` is never called recursively (using `ThreadPool.assertCurrentMethodIsNotCalledRecursively`)?

Reply:
This can't be done without some refactoring since tests use `Runnable::run` as the executor, which would of course execute on the same thread (recursively).